fvm/evm/emulator/emulator.go

```diff
@@ -0,0 +1,306 @@
+	res, err := proc.run(msg, tx.Type())
+	if err != nil {
+		return res, err
+	}
+
+	return res, bl.commit(res.StateRootHash)
+}
+
+func (bl *BlockView) newProcedure() (*procedure, error) {
+	execState, err := newState(bl.database)
+	if err != nil {
+		return nil, err
+	}
+	cfg := bl.config
+	return &procedure{
+		config: cfg,
+		evm: gethVM.NewEVM(
+			*cfg.BlockContext,
+			*cfg.TxContext,
+			execState,
+			cfg.ChainConfig,
+			cfg.EVMConfig,
+		),
+		state: execState,
+	}, nil
+}
+
+func (bl *BlockView) commit(rootHash gethCommon.Hash) error {
+	// commit atree changes back to the backend
+	err := bl.database.Commit(rootHash)
+	return handleCommitError(err)
+}
+
+type procedure struct {
+	config *Config
+	evm    *gethVM.EVM
+	state  *gethState.StateDB
+}
+
+// commit commits the changes to the state.
+func (proc *procedure) commit() (gethCommon.Hash, error) {
+	// commits the changes from the journal into the in memory trie.
+	// in the future if we want to move this to the block level we could use finalize
+	// to get the root hash
+	newRoot, err := proc.state.Commit(true)
+	if err != nil {
+		return gethTypes.EmptyRootHash, handleCommitError(err)
+	}
+
+	// flush the trie to the lower level db
+	// the reason we have to do this, is the original database
+	// is designed to keep changes in memory until the state.Commit
+	// is called then the changes moves into the trie, but the trie
+	// would stay in memory for faster transaction execution. you
+	// have to explicitly ask the trie to commit to the underlying storage
+	err = proc.state.Database().TrieDB().Commit(newRoot, false)
+	if err != nil {
+		return gethTypes.EmptyRootHash, handleCommitError(err)
+	}
+	return newRoot, nil
+}
+
+func handleCommitError(err error) error {
+	if err == nil {
+		return nil
+	}
+	// if known types (database errors) don't do anything and return
+	if types.IsAFatalError(err) || types.IsADatabaseError(err) {
+		return err
+	}
+
+	// else is a new fatal error
+	return types.NewFatalError(err)
+}
+
+func (proc *procedure) mintTo(address types.Address, amount *big.Int) (*types.Result, error) {
+	var err error
+	addr := address.ToCommon()
+	res := &types.Result{
+		GasConsumed: proc.config.DirectCallBaseGasUsage,
+		TxType:      types.DirectCallTxType,
+	}
+
+	// create account if not exist
+	if !proc.state.Exist(addr) {
+		proc.state.CreateAccount(addr)
+	}
+
+	// add balance
+	proc.state.AddBalance(addr, amount)
+
+	// we don't need to increment any nonce, given the origin doesn't exist
+	res.StateRootHash, err = proc.commit()
+
+	return res, err
+}
+
+func (proc *procedure) withdrawFrom(address types.Address, amount *big.Int) (*types.Result, error) {
+	var err error
+
+	addr := address.ToCommon()
+	res := &types.Result{
+		GasConsumed: proc.config.DirectCallBaseGasUsage,
+		TxType:      types.DirectCallTxType,
+	}
+
+	// check source balance
+	// if balance is lower than amount return
+	if proc.state.GetBalance(addr).Cmp(amount) == -1 {
```

Comment:
I believe we have to check if account exists first? otherwise a 0 is returned as balance and we would subtract from non-existing account, which I'm not sure if it's intended?

Reply:
it's interesting, initially I thought we don't need to check existence here, given its only callable for bridged accounts but then I realized one might create a bridged account and never fund it or do anything with it, then call for withdrawal. updated at [26480ec](https://github.com/onflow/flow-go/pull/4927/commits/26480ec9950d85613a783d01301f9523d5fe9edf) to check.